use product config docs when feature flag is on

chenba opened the pull request.

chenba commented:
#### Notes to the reviewer

Much of the work for this patch was scouring the code base for usage of Stripe product and plan/price metadata and address each on a case by case basis.  I like to think that I did a thorough job of that, as well as following up on metadata related changes that were merged while I was working on this patch.  But it does make the review a bit more difficult, in terms of whether this patch truly and totally implements the needed changes.

chenba commented on packages/fxa-shared/test/subscriptions/configuration/helpers.ts:
```diff
@@ -1,377 +0,0 @@
-      }
-    });
-  });
-});
```
This file was combined into the neighboring `utils.ts`

chenba commented:
@StaberindeZA I created this draft PR with no tests in case you have a chance to take a look while I'm on PTO.  Note that for the back end changes, we'd like to fallback to product/plan metadata, unlike the clean branching that was done for Payments.  Thanks!

chenba commented:
Rebased main to resolve conflicts from the download url metadata rename patch and an IAP related change. 🤞

chenba commented:
@StaberindeZA marking this ready for review in case you get a chance to look at any of it tomorrow (Tues June 21).  I'll add some inline comment when I have the chance.  Thanks.

StaberindeZA reviewed (commented):
I just have the 1 comment at the moment.

I still have a bunch of testing I'd like to do locally, but at the moment once I enable `productConfigsFirestore`, before and/or after loading config into Firestore, I'm getting 500's from `/oauth/subscriptions/plans`. I had a quick look, and it looks like joi validations? Have you experienced something similar?

chenba commented on packages/fxa-auth-server/test/local/payments/subscription-reminders.js:
```diff
@@ -50,6 +50,7 @@ describe('SubscriptionReminders', () => {
   let mockStripeHelper;
   let reminder;
   let mockConfig;
+  let realDateNow;
```
The changes for this has nothing to do with the functionality related to metadata/product config.  There was an issue with `Date.now` not being restored after a test failure, leading to an OOM error elsewhere.

chenba commented:
> I still have a bunch of testing I'd like to do locally, but at the moment once I enable `productConfigsFirestore`, before and/or after loading config into Firestore, I'm getting 500's from `/oauth/subscriptions/plans`. I had a quick look, and it looks like joi validations? Have you experienced something similar?

I've updated the response validator to allow two types of objects in the plans list, and I also added extra filtering so that the plans in the list either has a `.configuration` or passes metadata validation.  (I still need to write/update tests for these changes.)

StaberindeZA commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -2784,16 +2787,61 @@ module.exports = function (log, config, bounces) {
+      ) {
+        // we are not using `determineLocale` because the product config might support more locales than the FxA supported locales list
+        const locales = message.acceptLanguage ? [message.acceptLanguage] : [];
+        const localizedConfigs = localizedPlanConfig(
```
When I tested this locally, for a valid locale, I wasn't getting the expected localizedConfigs back.

After some investigating, I found that `getMergedConfig` wasn't working cause `this.products` didn't have the data from Firestore in it, as expected.

This seems to be related to the `select()` in manager.ts `load` [method](https://github.com/mozilla/fxa/blob/380b990379e91ea24089ded224c8bb4c9a6cead4/packages/fxa-shared/payments/configuration/manager.ts#L77).

Once I removed the `select()` it worked as expected. Also just FYI, I tested this with Meghan last week, and she had the same result as me.

chenba commented:
@StaberindeZA  https://github.com/mozilla/fxa/pull/12947/commits/3dd5dde5667c5aed88bc07558dc64801b559fe51 is the new commit with changes based on a couple of your comments.  I still need to rebase on main and make changes on top of the changes from https://github.com/mozilla/fxa/pull/13377

StaberindeZA reviewed (approved):
LGTM! Thank you for the changes. Super happy to get this one in.
